spec/app/ics-028-cross-chain-validation/methods.md

```diff
@@ -654,40 +738,82 @@ function onChanOpenConfirm(
+    abortSystemUnless(gs.providerClientState != nil AND gs.providerClientState.Valid())
+    // - contains a valid providerConsensusState
+    abortSystemUnless(gs.providerConsensusState != nil AND gs.providerConsensusState.Valid())
+    // - contains an initial validator set that matches 
+    //   the validator set in the providerConsensusState (e.g., ICS 7)
+    abortSystemUnless(gs.initialValSet == gs.providerConsensusState.validatorSet)
+  }
+  if gs.distributionChannelId != "" {
+      // - if distributionChannelId is provided, it must the ID
+      //   of a channel connected to the "transfer" port
+      channelEnd = provableStore.get("channelEnds/ports/transfer/channels/{gs.distributionChannelId}")
+      abortSystemUnless(channelEnd != nil)
+  }
 
   // bind to ConsumerPortId port 
   err = portKeeper.bindPort(ConsumerPortId)
   // check whether the capability for the port can be claimed
   abortSystemUnless(err == nil)
 
-  // create client of the provider chain 
-  clientId = clientKeeper.CreateClient(gs.providerClientState, gs.providerConsensusState)
+  // set pre-CCV state
+  preCCV = gs.preCCV
+
+  if preCCV {
+    // start consumer chain in pre-CCV state;
+    // store the ID of the client of the provider chain
+    providerClient = connectionEnd.clientIdentifier
 
-  // store the ID of the client of the provider chain
-  providerClient = clientId
+    clientState = provableStore.get("clients/{connectionEnd.clientIdentifier}/clientState")
+  }
+  else {
+    // start consumer chain in normal CCV state;
+    // create client of the provider chain and store the ID
+    providerClient = clientKeeper.CreateClient(gs.providerClientState, gs.providerConsensusState)
+    
+    clientState = gs.providerClientState
+  }
 
   // compute (and store) the consumer unbonding period from the provider unbonding period
-  consumerUnbondingPeriod = ComputeConsumerUnbondingPeriod(gs.providerClientState.unbondingTime)
+  consumerUnbondingPeriod = ComputeConsumerUnbondingPeriod(clientState.unbondingTime)
 
   // set default value for HtoVSC
   HtoVSC[getCurrentHeight()] = 0
 
   // set the initial validator set for the consumer chain
   foreach val IN gs.initialValSet {
-    ccVal := CrossChainValidator{
-      address: hash(val.pubKey),
-      power: val.power
+    ccvValidatorSet[hash(val.pubKey)] = val
+  }
+
+  // set distribution channel ID
+  distributionChannelId = gs.distributionChannelId
+
+  // initiate handshake 
+  if preCCV {
+    // initiate CCV channel opening handshake
+    // i.e., use handleChanOpenInit as defined in ICS-26
+    datagram = ChanOpenInit{
+      order: ORDERED,
+      connectionHops: [gs.connId], // same as the CCV channel
+      portIdentifier: ConsumerPortId,
+      counterpartyPortIdentifier: ProviderPortId,
+      version: ccvVersion,
     }
-    validatorSet[ccVal.address] = ccVal
+    handleChanOpenInit(datagram)
+  }
+  else {
+    // initiate connection opening handshake
```

Comment:
I could use something like this:
```typescript
 // initiate connection opening handshake
    // i.e., use handleConnOpenInit as defined in ICS-26
    datagram = ConnOpenInit{
      clientIdentifier: providerClient,
      counterpartyClientIdentifier: // TODO
      version: "ccv"
    }
    connId = handleConnOpenInit(datagram)

    // initiate CCV channel opening handshake
    // i.e., use handleChanOpenInit as defined in ICS-26
    datagram = ChanOpenInit{
      order: ORDERED,
      connectionHops: [connId],
      portIdentifier: ConsumerPortId,
      counterpartyPortIdentifier: ProviderPortId,
      version: ccvVersion,
    }
    handleChanOpenInit(datagram)
```
But how do I get the `counterpartyClientIdentifier` needed in the `ConnOpenInit` datagram. Also, the `handle...` methods in ICS-26 need to be updated to return the IDs (see [implementation](https://github.com/cosmos/ibc-go/blob/ddf9bafda77e795afd8b345362c8c9ee96de76f1/modules/core/03-connection/keeper/handshake.go#L52)).

Reply:
ahh i see. The clientID is getting generated at the end of the proposal on the provider chain right? So theoretically it could be in genesis.

Though perhaps that's not worth worth the added effort. It does make life easier for relayers tho since they just automatically will complete the handshakes